File: pc/internal/modem/fft.go
package modem

import (
	"math"
	"math/cmplx"
)

// FFT computes the Discrete Fourier Transform using Cooley-Tukey radix-2.
// Input length must be a power of 2.
func FFT(x []complex128) []complex128 {
	n := len(x)
	if n <= 1 {
		out := make([]complex128, n)
		copy(out, x)
		return out
	}
	if n&(n-1) != 0 {
		panic("FFT: length must be a power of 2")
	}

	out := make([]complex128, n)
	copy(out, x)
	bitReverse(out)
	fftIterative(out, false)
	return out
}

// IFFT computes the Inverse Discrete Fourier Transform.
func IFFT(x []complex128) []complex128 {
	n := len(x)
	if n <= 1 {
		out := make([]complex128, n)
		copy(out, x)
		return out
	}

	out := make([]complex128, n)
	copy(out, x)
	bitReverse(out)
	fftIterative(out, true)

	// Scale by 1/N
	scale := 1.0 / float64(n)
	for i := range out {
		out[i] *= complex(scale, 0)
	}
	return out
}

func fftIterative(x []complex128, inverse bool) {
	n := len(x)
	for size := 2; size <= n; size <<= 1 {
		halfSize := size >> 1
		sign := -1.0
		if inverse {
			sign = 1.0
		}
		wn := cmplx.Exp(complex(0, sign*2*math.Pi/float64(size)))
		for start := 0; start < n; start += size {
			w := complex(1, 0)
			for j := 0; j < halfSize; j++ {
				u := x[start+j]
				v := w * x[start+j+halfSize]
				x[start+j] = u + v
				x[start+j+halfSize] = u - v
				w *= wn
			}
		}
	}
}

func bitReverse(x []complex128) {
	n := len(x)
	bits := 0
	for tmp := n; tmp > 1; tmp >>= 1 {
		bits++
	}
	for i := 0; i < n; i++ {
		j := reverseBits(i, bits)
		if i < j {
			x[i], x[j] = x[j], x[i]
		}
	}
}

func reverseBits(x, bits int) int {
	result := 0
	for i := 0; i < bits; i++ {
		result = (result << 1) | (x & 1)
		x >>= 1
	}
	return result
}

// RealFFT performs FFT on real-valued input.
func RealFFT(x []float64) []complex128 {
	n := len(x)
	cx := make([]complex128, n)
	for i, v := range x {
		cx[i] = complex(v, 0)
	}
	return FFT(cx)
}

// RealIFFT performs IFFT and returns only the real part.
func RealIFFT(x []complex128) []float64 {
	result := IFFT(x)
	out := make([]float64, len(result))
	for i, v := range result {
		out[i] = real(v)
	}
	return out
}

File: android/app/src/main/java/com/example/audiomodem/modem/FFT.kt
package com.example.audiomodem.modem

import kotlin.math.*

/**
 * FFT/IFFT implementation using Cooley-Tukey radix-2 algorithm.
 * Complex numbers are represented as pairs of DoubleArrays (real, imag).
 */
object FFT {

    /**
     * Compute FFT. Input length must be a power of 2.
     */
    fun fft(real: DoubleArray, imag: DoubleArray): Pair<DoubleArray, DoubleArray> {
        val n = real.size
        require(n > 0 && n and (n - 1) == 0) { "Length must be a power of 2" }

        val re = real.copyOf()
        val im = imag.copyOf()

        bitReverse(re, im)
        fftIterative(re, im, false)

        return Pair(re, im)
    }

    /**
     * Compute IFFT. Input length must be a power of 2.
     */
    fun ifft(real: DoubleArray, imag: DoubleArray): Pair<DoubleArray, DoubleArray> {
        val n = real.size
        require(n > 0 && n and (n - 1) == 0) { "Length must be a power of 2" }

        val re = real.copyOf()
        val im = imag.copyOf()

        bitReverse(re, im)
        fftIterative(re, im, true)

        // Scale by 1/N
        val scale = 1.0 / n
        for (i in 0 until n) {
            re[i] *= scale
            im[i] *= scale
        }

        return Pair(re, im)
    }

    /**
     * FFT of real-valued input.
     */
    fun realFft(x: DoubleArray): Pair<DoubleArray, DoubleArray> {
        return fft(x, DoubleArray(x.size))
    }

    /**
     * IFFT returning only the real part.
     */
    fun realIfft(real: DoubleArray, imag: DoubleArray): DoubleArray {
        val (re, _) = ifft(real, imag)
        return re
    }

    private fun fftIterative(re: DoubleArray, im: DoubleArray, inverse: Boolean) {
        val n = re.size
        var size = 2
        while (size <= n) {
            val halfSize = size / 2
            val sign = if (inverse) 1.0 else -1.0
            val angle = sign * 2.0 * PI / size

            var start = 0
            while (start < n) {
                var wRe = 1.0
                var wIm = 0.0
                val wnRe = cos(angle)
                val wnIm = sin(angle)

                for (j in 0 until halfSize) {
                    val idx1 = start + j
                    val idx2 = start + j + halfSize

                    val tRe = wRe * re[idx2] - wIm * im[idx2]
                    val tIm = wRe * im[idx2] + wIm * re[idx2]

                    re[idx2] = re[idx1] - tRe
                    im[idx2] = im[idx1] - tIm
                    re[idx1] += tRe
                    im[idx1] += tIm

                    val newWRe = wRe * wnRe - wIm * wnIm
                    val newWIm = wRe * wnIm + wIm * wnRe
                    wRe = newWRe
                    wIm = newWIm
                }

                start += size
            }
            size *= 2
        }
    }

    private fun bitReverse(re: DoubleArray, im: DoubleArray) {
        val n = re.size
        var bits = 0
        var tmp = n
        while (tmp > 1) {
            bits++
            tmp = tmp shr 1
        }

        for (i in 0 until n) {
            val j = reverseBits(i, bits)
            if (i < j) {
                var temp = re[i]; re[i] = re[j]; re[j] = temp
                temp = im[i]; im[i] = im[j]; im[j] = temp
            }
        }
    }

    private fun reverseBits(x: Int, bits: Int): Int {
        var result = 0
        var v = x
        for (i in 0 until bits) {
            result = (result shl 1) or (v and 1)
            v = v shr 1
        }
        return result
    }
}
